wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java

```diff
@@ -0,0 +1,304 @@
+import org.eclipse.che.dto.server.DtoFactory;
+
+/** Defines Devfile REST API. */
+@Api(value = "/devfile", description = "Devfile REST API")
+@Path("/devfile")
+public class DevfileService extends Service {
+
+  private final DevfileSchemaProvider schemaCachedProvider;
+  private final UserDevfileManager userDevfileManager;
+  private final DevfileServiceLinksInjector linksInjector;
+
+  @Inject
+  public DevfileService(
+      DevfileSchemaProvider schemaCachedProvider,
+      UserDevfileManager userDevfileManager,
+      DevfileServiceLinksInjector linksInjector) {
+    this.userDevfileManager = userDevfileManager;
+    this.linksInjector = linksInjector;
+    this.schemaCachedProvider = schemaCachedProvider;
+  }
+
+  /**
+   * Retrieves the json schema.
+   *
+   * @return json schema
+   */
+  @GET
+  @Produces(APPLICATION_JSON)
+  @ApiOperation(value = "Retrieves current version of devfile JSON schema")
+  @ApiResponses({
+    @ApiResponse(code = 200, message = "The schema successfully retrieved"),
+    @ApiResponse(code = 500, message = "Internal server error occurred")
+  })
+  public Response getSchema() throws ServerException {
+    try {
+      return Response.ok(schemaCachedProvider.getSchemaContent(CURRENT_API_VERSION)).build();
+    } catch (IOException e) {
+      throw new ServerException(e);
+    }
+  }
+
+  @Path("/devfile")
+  @POST
+  @Consumes({APPLICATION_JSON, "text/yaml", "text/x-yaml"})
+  @Produces(APPLICATION_JSON)
+  @ApiOperation(
+      value = "Creates a new persistent Devfile from yaml representation",
+      consumes = "application/json, text/yaml, text/x-yaml",
+      produces = APPLICATION_JSON,
+      nickname = "createFromDevfileYaml",
+      response = UserDevfileDto.class)
+  @ApiResponses({
+    @ApiResponse(code = 201, message = "The devfile successfully created"),
+    @ApiResponse(code = 400, message = "Missed required parameters, parameters are not valid"),
+    @ApiResponse(code = 403, message = "The user does not have access to create a new devfile"),
+    @ApiResponse(code = 409, message = "Conflict error occurred during the devfile creation"),
+    @ApiResponse(code = 500, message = "Internal server error occurred")
+  })
+  public Response createFromDevfileYaml(
+      @ApiParam(value = "The devfile to create", required = true) DevfileDto devfile)
+      throws ConflictException, BadRequestException, ForbiddenException, NotFoundException,
+          ServerException {
+    requiredNotNull(devfile, "Devfile");
+    return Response.status(201)
+        .entity(
+            linksInjector.injectLinks(
+                asDto(
+                    userDevfileManager.createDevfile(
+                        DtoFactory.newDto(UserDevfileDto.class)
+                            .withDevfile(devfile)
+                            .withName(NameGenerator.generate("devfile-", 16)))),
+                getServiceContext()))
+        .build();
+  }
+
+  @POST
+  @Consumes({APPLICATION_JSON})
```

Comment:
why not accept yaml here?

Reply:
I see no use-case for that. Do you?

Reply:
I don't know. Hard to imagine all use-cases. I guess we can add it in the future, if need raises